Detect rustc(drivers) more precisely

sagudev opened the pull request:
Fixes https://github.com/mozilla/sccache/issues/861

sagudev commented:
Aarch fails on unrelated error (installation of musl).

Xuanwo commented:
> Aarch fails on unrelated error (installation of musl).

Yep. It's unrelated. I'm working on the fix. Please ignore it for now. Thanks for the patience.

sylvestre commented on src/compiler/compiler.rs:
```diff
@@ -921,129 +944,137 @@ where
-                if stdout.starts_with("rustc ") {
-                    return Ok(stdout);
-                }
+    let rustc_vv = run_input_output(child, None).await.and_then(|output| {
```
maybe move this block into a function ? 
it is becoming a bit too long

sylvestre commented on src/compiler/compiler.rs:
```diff
@@ -921,129 +944,137 @@ where
-            Ok(rustc_verbose_version) => {
-                let rustc_executable2 = rustc_executable.clone();
+    // If it does not look like rustc and it's verification failed check for c compiler
+    match rustc_vv {
```
same, could you please try to move this into a function

sagudev replied:
I moved both into one function as both are tightly coupled.

sylvestre commented on src/compiler/compiler.rs:
```diff
@@ -921,129 +944,174 @@ where
+
+    // Even if it does not look like rustc like it might still be rustc driver
+    // so we do full check
+    match verified_rust_compiler(
```
if I understand correctly, this function is performing an action

So, it should probably called
verify_rust_compiler

sylvestre commented on src/compiler/compiler.rs:
```diff
@@ -921,129 +944,174 @@ where
+
+#[allow(clippy::too_many_arguments)]
+/// If `must_be_rust` is `true` and verification of `rustc_executable` fails it returns error,
+/// if `must_be_rust` is `false` returns `None` if verification failed or `Some((Compiler, CompilerProxy))` if it succeeded.
```
it would be nice in this comment to explicit what can be must_be_rust=false

drahnr commented on src/compiler/compiler.rs:
```diff
@@ -921,129 +944,176 @@ where
                     (
-                        Box::new(c) as Box<dyn Compiler<T>>,
-                        proxy as Option<Box<dyn CompilerProxy<T>>>,
+                        proxy
+                            .map(Box::new)
+                            .map(|x: Box<RustupProxy>| x as Box<dyn CompilerProxy<T>>),
+                        resolved_compiler_executable,
```
This became less readable than before

sagudev replied:
Are you doing per commit review as I cannot find this comment on full view?

drahnr replied:
I am doing a full diff view, and the changeset exists.

drahnr commented on src/compiler/compiler.rs:
```diff
@@ -907,6 +907,29 @@ fn is_rustc_like<P: AsRef<Path>>(p: P) -> bool {
+            .file_stem()
+            .map(|s| s.to_string_lossy().to_lowercase())
+            .as_deref(),
+        Some("cc")
```
```suggestion
        Some("cc" | "c++" | .. )
```
should work as well

drahnr commented on src/compiler/compiler.rs:
```diff
@@ -921,129 +944,176 @@ where
     T: CommandCreatorSync,
 {
     trace!("detect_compiler: {}", executable.display());
-
     // First, see if this looks like rustc.
 
-    let rustc_executable = if is_rustc_like(executable) {
-        Some(executable.to_path_buf())
+    let mut must_be_rust = false;
+    let mut rustc_executable = executable.to_path_buf();
+    if is_rustc_like(executable) {
+        must_be_rust = true;
     } else if env.iter().any(|(k, _)| k == OsStr::new("CARGO")) {
         // If not, detect the scenario where cargo is configured to wrap rustc with something other than sccache.
         // This happens when sccache is used as a `RUSTC_WRAPPER` and another tool is used as a
         // `RUSTC_WORKSPACE_WRAPPER`. In that case rustc will be the first argument rather than the command.
         //
         // The check for the `CARGO` env acts as a guardrail against false positives.
         // https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-reads
-        args.iter().next().and_then(|arg1| {
+        if let Some(arg1) = args.iter().next() {
             if is_rustc_like(arg1) {
-                Some(PathBuf::from(arg1))
-            } else {
-                None
+                must_be_rust = true;
+                rustc_executable = PathBuf::from(arg1)
             }
-        })
-    } else {
-        None
+        }
```
I'd prefer to use a local `struct X {}` and avoid too many `mut` variables floating around in a scope.

drahnr commented:
I have yet to do a full review.

sagudev commented:
I squashed all commits as they were not meant to be reviewed separately. I also updated `verify_rust_compiler` to not depend on `must_be_rust` and moved this logic back to `detect_compiler`

drahnr commented on src/compiler/compiler.rs:
```diff
@@ -921,129 +938,166 @@ where
-        match rustc_vv {
-            Ok(rustc_verbose_version) => {
-                let rustc_executable2 = rustc_executable.clone();
+/// Tries to verify that provided executable is really rust compiler
+/// or rust driver
+async fn verify_rust_compiler<T>(
```
Requires more documentation on the disriminator used and the function name needs a better name.

sagudev replied:
What is wrong with name? Main part of this function is:
```rust
// rustc -vV verification status
match rustc_vv {
   ...
```
anything specific you have in mind?

I also do not think there is much to document about discriminator used as there is no discriminator used. We check for rust compiler using -vV instead of using binary name and then verifying compiler with -vV.

drahnr replied:
It does attempt to _resolve_ a given path to the true (`rustc`) compiler binary with a failure mode if it cannot, so verification is a red hering and doesn't describe what it is doing. That's my main issue

sagudev replied:
What about `resolve_rust_compiler`?

sagudev commented:
rebased

drahnr commented:
I'll do a full review as soon as all CI steps went green

sagudev commented:
Checks are green in my fork: https://github.com/sagudev/sccache/actions/runs/7656273169 & https://github.com/sagudev/sccache/actions/runs/7656273174

drahnr commented on src/compiler/compiler.rs:
```diff
@@ -980,6 +980,23 @@ fn is_rustc_like<P: AsRef<Path>>(p: P) -> bool {
+            .file_stem()
+            .map(|s| s.to_string_lossy().to_lowercase())
+            .as_deref(),
+        Some("cc" | "c++")
+            | Some("gcc" | "g++")
+            | Some("clang" | "clang++")
+            | Some("clang-cl" | "cl")
+            | Some("nvc" | "nvc++" | "nvcc")
+    )
```
```suggestion
        Some("cc" | "c++"
            | "gcc" | "g++"
            | "clang" | "clang++"
            | "clang-cl" | "cl"
            | "nvc" | "nvc++" | "nvcc")
    )
```

sagudev commented on src/compiler/compiler.rs:
```diff
@@ -994,129 +1011,166 @@ where
-                if stdout.starts_with("rustc ") {
-                    return Ok(stdout);
-                }
+    if !must_be_rust && is_like_c_compiler(&rustc_executable) {
```
This is to be compatible with current behavior. If name of compiler is rustc and we cannot detect it we fast fail (we do not try it for c compiler).

drahnr replied:
That may be, but it can be expressed more concisely.

sagudev replied:
So, should I just change behavior and simply code or express it differently (currently have no idea how)?

drahnr commented on src/compiler/compiler.rs:
```diff
@@ -1399,6 +1453,14 @@ mod test {
+                "",
+                "gcc: error: unrecognized command-line option '-vV'; did you mean '-v'?",
+            )),
+        );
```
Iirc, the result will be cached, so the impact of another resolution call is very small, even for small projects. Correct me if I am wrong.

sagudev replied:
Is this comment to https://github.com/mozilla/sccache/pull/1897#discussion_r1466719055?

drahnr replied:
This is a principled question, if we incur much overhead for the compiler resolution with this changeset.

sagudev replied:
I am not sure, to be honest. But if the results are cached then it `should` be ok.

drahnr reviewed (changes requested):
I don't particularly like the use mutable `bool` variables in deciding the compiler. There is an outstanding question on the efficiency of test runs and ifwe can assume `-vV` is unique to rust.

sagudev commented:
>  if we can assume `-vV` is unique to rust.

Even if it is not we try to parse the result and if it does not contain rustc(-drivers) specific output we know it is not rusty.

sylvestre commented:
moved into a different PR
